### PR TITLE
feat: allow runtime specializations rather than application state

### PR DESCRIPTION
**Pull request by gsdevme**
Removes state held by Needle and any application state driven concepts in favour of context passing to allow stateless APIs, mutations inline with expectations of an application consuming Needle.

### CODE REVIEW

File: Attribute/AttributeProviderInterface.php [hunk 1]
```diff
@@ -11,5 +11,5 @@ interface AttributeProviderInterface
     public function getAttributeByName(
         string $name,
         SpecializationContextHashInterface $contextHash
-    ): ?AttributeInterface;
+    ): AttributeInterface;
```
Comment (gsdevme) [inline Attribute/AttributeProviderInterface.php:14]: What happens if the attribute doesn’t exist now? @calumbrodie  this change hasn’t introduced an exception or how to handle that?

Comment (gsdevme) [reply Attribute/AttributeProviderInterface.php:14]: You might also find this is fairly breaky, as this code has been null able for years

Comment (calumbrodie) [reply Attribute/AttributeProviderInterface.php:14]: These providers always return a generic attribute anyway so it makes consuming code not have to check for null returns

**Comment by calumbrodie** [comment]
Suggest we move to 1.0.0 after merging this (this is a major version)

**Pull request merged**